lib/Common/Codex/Utf8Codex.cpp

```diff
@@ -395,10 +394,11 @@ namespace utf8
             cch -= 4;
         }
 LSlowPath:
+        // WARNING: Assume cch correct, suppress end-of-buffer checking
+        // we might have failed to measure cch, +1 for odd length. sizeof(char16) == 2
+        LPCUTF8 end = ptr + cch + (cch % 2 ? 2 : 1);
         while (cch-- > 0)
         {
-            LPCUTF8 end = ptr + cch + 1; // WARNING: Assume cch correct, suppress end-of-buffer checking
```

Comment:
cc @boingoing 

I wonder if both before and after change are incorrect. We don't know "end" at all based on "cch" (number of characters). The original code before Taylor's change makes some sense: "*buffer++ = Decode(ptr, ptr + 4, localOptions)", because a character can map up to 3 utf8 code units. Now either in-loop "ptr + cch + 1" (assume input cch == 1, end == ptr + 1, we can only decode 1 byte instead of up to 3) or out-of-loop "ptr + cch + 2/1" (assume input cch == 10, 20-byte utf8 stream, end == ptr + 12/11 instead of ptr + 20) could cut the input stream to much shorter and produce wrong result. How can this work?